### PR TITLE
Memory tagging

### PR DESCRIPTION
**Pull request by jenswi-linaro**

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Depends on https://github.com/OP-TEE/optee_os/pull/5276

**Comment by jenswi-linaro** [comment]
Another option is to use the `handle.h` approach, like in pkcs11_ta.

**Comment by jforissier** [comment]
> Another option is to use the `handle.h` approach, like in pkcs11_ta.

Ah yes, that would be less "destructive" and quite clean indeed.

**Comment by jenswi-linaro** [comment]
OK, I'll update.

**Comment by jenswi-linaro** [comment]
Updated

File: ta/os_test/os_test.c [hunk 1]
```diff
@@ -1460,3 +1462,58 @@ TEE_Result ta_entry_dl_phdr_dl(void)
...
+	*p = 32;
+
+	p2 = memtag_insert_tag(p, (memtag_get_tag(p) + 1) & 0xf);
+	*p2 = 42;
+	EMSG("Expected to crash when writing to %p which was allocated as %p",
+	     (void *)p2, (void *)p);
+err:
+	TEE_Free(p);
+	return TEE_ERROR_GENERIC;
+}
```
Comment (jforissier) [inline ta/os_test/os_test.c:1503]: Could you also add a test for invalid tag due to some basic heap overflow without manipulating the tag?

File: ta/crypt/cryp_taf.c [hunk 2]
```diff
@@ -2,21 +2,110 @@
...
+static uint32_t op_handle_get(TEE_OperationHandle h)
+{
+	int rc = handle_get(&op_db, h);
+
+	if (rc < 0) {
+		EMSG("Failed to allocate handle");
+		TEE_Panic(TEE_ERROR_OUT_OF_MEMORY);
+	}
+
+	return rc + 1;
```
Comment (jforissier) [inline ta/crypt/cryp_taf.c:35]: Add `/* 0 (TEE_HANDLE_NULL) is invalid */` perhaps?

**Review by jforissier — COMMENTED** [review]
"ta/crypt: update handle conversion to support MTE":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jenswi-linaro** [comment]
Tag applied, updated the 1038 test as requested.

**Comment by jforissier** [comment]
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

Thanks!

**Comment by jenswi-linaro** [comment]
Tag applied

**Comment by jforissier** [comment]
Not strictly related to the code in this PR but with `MEMTAG=y` there is a assertion failure in GP test 20011 and a couple of others:
```
E/TA:  assertion 'memtag_get_tag(addr) == memtag_read_tag(addr)' failed at lib/libutils/ext/include/memtag.h:214 in memtag_assert_tag()
E/TC:? 0
E/TC:? 0 TA panicked with code 0xffff0000
E/LD:  Status of TA 534d4152-542d-4353-4c54-2d54412d3031
E/LD:   arch: aarch64
E/LD:  region  0: va 0x40004000 pa 0x0e326000 size 0x002000 flags rw-s (ldelf)
E/LD:  region  1: va 0x40006000 pa 0x0e328000 size 0x008000 flags r-xs (ldelf)
E/LD:  region  2: va 0x4000e000 pa 0x0e330000 size 0x001000 flags rw-s (ldelf)
E/LD:  region  3: va 0x4000f000 pa 0x0e331000 size 0x004000 flags rw-s (ldelf)
E/LD:  region  4: va 0x40013000 pa 0x0e335000 size 0x001000 flags r--s
E/LD:  region  5: va 0x40014000 pa 0x0e359000 size 0x001000 flags rw-s (stack)
E/LD:  region  6: va 0x4001b000 pa 0x00001000 size 0x017000 flags r-xs [0] .ta_head .text .eh_frame .rodata .gnu.hash .dynsym .dynstr .hash .rela.dyn
E/LD:  region  7: va 0x40032000 pa 0x00018000 size 0x00c000 flags rw-s [0] .dynamic .got .rela.got .data .bss
E/LD:   [0] 534d4152-542d-4353-4c54-2d54412d3031 @ 0x4001b000 (out-br/build/optee_test_ext-1.0/host/xtest/gp-suite/TTAs_Internal_API_1_1_1/TTA_TCF/TTA_TCF/code_files/out/534d4152-542d-4353-4c54-2d54412d3031.elf)
E/LD:  Call stack:
E/LD:   0x4002292c _assert_break at optee_os/lib/libutee/assert.c:24
E/LD:   0x40026b14 memtag_assert_tag at optee_os/lib/libutils/ext/include/memtag.h:214
E/LD:   0x400277e8 raw_free at optee_os/lib/libutils/isoc/bget_malloc.c:454
E/LD:   0x4001c2f0 CmdTEEFreePropertyEnumerator at out-br/build/optee_test_ext-1.0/host/xtest/gp-suite/TTAs_Internal_API_1_1_1/TTA_TCF/TTA_TCF/code_files/TTA_TCF.c:867
E/LD:   0x400209a0 entry_invoke_command at optee_os/lib/libutee/arch/arm/user_ta_entry.c:370
E/LD:   0x4001d900 __ta_entry at optee_os/out/arm/export-ta_arm64/src/user_ta_header.c:48
```

It seems the test code is buggy and a similar fix to "ta/crypt: update handle conversion to support MTE" would be needed.

**Comment by jforissier** [comment]
I should open a new issue for the above asserts, this PR is good for merging.
